### PR TITLE
Add basic functionality for validation set

### LINKED ISSUE #60: Logging of testing leads to invalid json files.
The json output file contains two root elements (one for train and one for testing) and is therefore not valid. For this reason, we can no longer use ```json.loads()``` to load the files (but there are [workarounds](https://stackoverflow.com/questions/21058935/python-json-loads-shows-valueerror-extra-data/51830719)).

I suggest to either create two output files (one for training, one for testing) or to integrate the testing values as an object into training. 

The solution should also take into account how the validation will/should be logged.

### LINKED ISSUE #61: Enable to also generate validation set
Currently data is prepared for train and test. We should therefore also support the preparation of a validation set.

### PR DESCRIPTION
**Pull request by eileen-kuehn**
This PR adds basic functionality to also generate a validation set for the mnist dataset.
Further, this PR improves the data loading by removing duplicates from the different sets of train, validation and test sets.
I also changed some of the methods to improve the speed of data loading.

And one more thing: I did hide the writing of newlines after each json object while logging the results.

Closes #61, closes #60.

### CODE REVIEW

File: maskit/datasets/mnist.py [hunk 1]
```diff
@@ -48,51 +30,95 @@ def apply_PCA(wires: int, x_train: np.ndarray):
...
-
-    x_train, y_train = x_train[:train_size], y_train[:train_size]
-    x_test, y_test = x_test[:test_size], y_test[:test_size]
-
-    x_train = convert_to_binary(x_train)
-    x_test = convert_to_binary(x_test)
-    y_train = [convert_label(y, classes) for y in y_train]
-    y_test = [convert_label(y, classes) for y in y_test]
+    if shuffle:
+        x_train, y_train = skl_shuffle(x_train, y_train)
```
Comment (cDenius) [inline maskit/datasets/mnist.py:88]: ```skl_shuffle``` has a ```random_state``` parameter. Consequently, in the current form it should be returning a different shuffle for every run and therefore result in a different validation set. I think we should set the parameter to a fixed number.

Comment (eileen-kuehn) [reply maskit/datasets/mnist.py:88]: Hm, I thought it might sense to not set it at the moment. If we set it to a fixed number, we would only have two possibilities, either shuffling or no shuffling. If we don't set it, we will get a real shuffled dataset as expected. So we would have to define another seed for data to make it more generic. This might be something better suited for the future?

Comment (cDenius) [reply maskit/datasets/mnist.py:88]: But if we have a different validation set every run, we cant use its accuracy to compare the influence of different hyperparameter values. 
E.g. If you choose a step_size of 0.01 and get a validation score of 0.9 and with step_size 0.001 a validation score of 0.95, you cant be sure if it is because of the step_size or because of the different validation set. You would have to wait for the test set results at the end of the training.

Comment (eileen-kuehn) [reply maskit/datasets/mnist.py:88]: Wouldn't you use `shuffle=False` in those cases? This always produces the same datasets for training and validation. Maybe I don't get your use case correctly O:)

Comment (cDenius) [reply maskit/datasets/mnist.py:88]: Oh yeah, i see. I missed again that the train dataset is already shuffled. Then it should be alright.

Comment (maxfischer2781) [inline maskit/datasets/mnist.py]: *Iff* we also accept ``shuffle`` as an int, we could directly use `sklearn.utils.resample` instead here.

File: maskit/datasets/mnist.py [hunk 2]
```diff
@@ -48,51 +30,95 @@ def apply_PCA(wires: int, x_train: np.ndarray):
...
+
+    # split validation set from train set
+    split_point = len(x_train) - (5000 if validation_size > 0 else 0)
```
Comment (maxfischer2781) [inline maskit/datasets/mnist.py:91]: Is there a reason why this is always 5000 and not the validation_size?

**Review by maxfischer2781 — CHANGES_REQUESTED** [review]
Bit late to the party…

File: maskit/datasets/mnist.py (hunk 2 above)
Comment (eileen-kuehn) [reply maskit/datasets/mnist.py:91]: Yes, indeed! 🤓
I borrowed this number based on a Tensorflow's MNIST tutorial that used 5,000 images as a validation set. However, there is also a paper on *What is the Best Multi-Stage Architecture for Object Recognition?*, in which they use a set of 10,000 validation images. So I am glad to change this if you have a specific opinion on this.

File: maskit/datasets/mnist.py (hunk 1 above)
Comment (eileen-kuehn) [reply maskit/datasets/mnist.py]: I also introduced a seed – should be future proof now 👍 Great hint, now we are even more flexible 🤪.

File: maskit/datasets/mnist.py (hunk 2 above)
Comment (cDenius) [reply maskit/datasets/mnist.py:91]: You could do the splitting of training and validation set after calling ```prepare_data()``` with ```train_size + validation_size``` instead of calling it for each one individually.

But then you have to decide what to do if there are not enough data points for both of them. Reduce the size of the training set or of the validation set?

Comment (eileen-kuehn) [reply maskit/datasets/mnist.py:91]: I just created a ticket out of this discussion (see #63) so we can postpone any decisions.

File: main.py [hunk 3]
```diff
@@ -197,72 +225,112 @@ def cost_fn(params, masked_circuit=None):
...
+    masked_circuit: MaskedCircuit,
+    rotations: List,
+    wires_to_measure: Tuple[int, ...],
+    interpret: Tuple[int, ...],
+    data: Optional[np.ndarray] = None,
+    target: Optional[np.ndarray] = None,
+):
+    if data is None or target is None:
+        pass
+    elif data is not None and target is not None:
```
Comment (maxfischer2781) [inline main.py:323]: The ``pass`` case is already excluded by this one. You can remove it.
```suggestion
    if data is not None and target is not None:
```

**Review by maxfischer2781 — APPROVED** [review]
Looks good to me. Only found one thing in ``main.py`` that could be condensed – don't midn either way, though.

**Review by maxfischer2781 — APPROVED** [review]
Make it so!
